### PR TITLE
Complete documentation for 3.11.0

### PR DESCRIPTION
**Pull request by nlohmann**
- BJData
- contributors for 3.11.0

Further changes:

- removed deprecated BJData functions
- removed example for deprecated `iterator_wrapper` function
- added a lot more examples

Todo:

- [x] check descriptions of exceptions (see https://github.com/nlohmann/json/pull/3463#discussion_r862485440)
- [x] check binary size table (see https://github.com/nlohmann/json/pull/3336#issuecomment-1114333114)
- [x] have @fangq have a look at the documentation for BJData

### CODE REVIEW

File: docs/mkdocs/docs/home/exceptions.md [hunk 1]
```diff
@@ -295,13 +295,28 @@ When parsing CBOR or MessagePack, the byte vector ends before the complete value
 
 ### json.exception.parse_error.112
 
-Not all types of CBOR or MessagePack are supported. This exception occurs if an unsupported byte was read.
+An unexpected byte was read in a [binary format](../features/binary_formats/index.md) or length information is invalid ([BSON](../features/binary_formats/bson.md)).
 
 !!! failure "Example message"
```
Comment (falbrechtskirchinger) [inline docs/mkdocs/docs/home/exceptions.md:300]: "messages"

**Review by falbrechtskirchinger — COMMENTED** [review]
:+1:

File: docs/mkdocs/docs/home/exceptions.md (hunk 1 above)
Comment (nlohmann) [reply docs/mkdocs/docs/home/exceptions.md]: Done in aacd97a1c0190657db231707a2e2be2ac5f02f57.

File: docs/examples/object_comparator_t.cpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#include <iostream>
+#include <nlohmann/json.hpp>
+
+using json = nlohmann::json;
+
+int main()
+{
+    std::cout << std::boolalpha
+              << "one < two : " << json::object_comparator_t{}("one", "two") << "\n"
+              << "three < four : " << json::object_comparator_t{}("three", "four") << std::endl;
```
Comment (falbrechtskirchinger) [inline docs/examples/object_comparator_t.cpp:10]: `object_comparator_t` is not guaranteed to be `std::less` anymore. Could also be  `std::equal_to` or possibly something else depending on `object_t::key_compare`.

Comment (nlohmann) [reply docs/examples/object_comparator_t.cpp:10]: What about

```cpp
std::cout << std::boolalpha
          << "json::object_comparator_t{}(\"one\", \"two\") = " << json::object_comparator_t{}("one", "two") << "\n"
          << "json::object_comparator_t{}(\"three\", \"four\") = : " << json::object_comparator_t{}("three", "four") << std::endl;
```

?

Comment (falbrechtskirchinger) [reply docs/examples/object_comparator_t.cpp:10]: Sure. (Without the erroneous ':' in `\"four\") = : "`.)

Comment (nlohmann) [reply docs/examples/object_comparator_t.cpp]: Done. (The code above never saw an IDE ;-))

File: docs/examples/~basic_json.cpp [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+using json = nlohmann::json;
+
+int main()
+{
+    json* j = new json("Hello, world!");
+    std::cout << *j << std::endl;
+
+    // explicitly call destructor
+    delete j;
```
[the remark this answers is not in the document]
Comment (falbrechtskirchinger) [reply docs/examples/~basic_json.cpp:12]:
```cpp
{
    json j("Hello, world!");
} // j goes out of scope, which calls the destructor
```

My point is, that neither one is a useful example. Using heap allocation doesn't add anything IMHO.

If you're explicitly calling the destructor, at least the documented function appears in the actual source code.

[the remark this answers is not in the document]
Comment (falbrechtskirchinger) [reply docs/examples/~basic_json.cpp:12]: @nlohmann Am I correct in assuming you're just trying to satisfy the linter? Maybe it's time to add `<!-- NOLINT(rule) -->`.

[the remark this answers is not in the document]
Comment (nlohmann) [reply docs/examples/~basic_json.cpp:12]: I just wanted to give an example to everything in the public API.

What about the example from `get_allocator`:

```cpp
auto alloc = json::get_allocator();
using traits_t = std::allocator_traits<decltype(alloc)>;

json* j = traits_t::allocate(alloc, 1);
traits_t::construct(alloc, j, "Hello, world!");

std::cout << *j << std::endl;

// call destructor
traits_t::destroy(alloc, j);
traits_t::deallocate(alloc, j, 1);
```

But I'm also fine with @gregmarr 's proposal.

[the remark this answers is not in the document]
Comment (falbrechtskirchinger) [reply docs/examples/~basic_json.cpp:12]: > ```
> json j("Hello, world!");
> std::cout << std::boolalpha << "is string? " << j.is_string() << "\n";
> j.~json();
> std::cout <<  "is discarded? " << j.is_discarded() << "\n";
> ```
> 
> That's not legal. You are using `j` after it was destroyed.

Is this UB according to the standard? The memory is still there after all. I'll have to investigate that later, I'm curious now.
Doing a placement new doesn't show the destructor action either. But at least it involves an explicit destructor call.

> I just wanted to give an example to everything in the public API.
> 
> What about the example from `get_allocator`:

Again, the destructor call isn't *visible*. I find it all very unsatisfying and not particularly illustrative.

[the remark this answers is not in the document]
Comment (falbrechtskirchinger) [reply docs/examples/~basic_json.cpp:12]: > Once a destructor is invoked for an object, the object's lifetime ends; the behavior is undefined if the destructor is invoked for an object whose lifetime has ended ([basic.life]).
Fair enough,

[the remark this answers is not in the document]
Comment (nlohmann) [reply docs/examples/~basic_json.cpp]: I removed the example.

**Comment by fangq** [comment]
@nlohmann, let me know if the above commit looks ok to you: https://github.com/NeuroJSON/json/commit/d2238e58f27cb3a1524c8432c04410ff4c0e25d3
or
https://github.com/NeuroJSON/json/blob/issue3464/docs/mkdocs/docs/features/binary_formats/bjdata.md

**Comment by falbrechtskirchinger** [comment]
> @nlohmann, let me know if the above commit looks ok to you: [NeuroJSON@d2238e5](https://github.com/NeuroJSON/json/commit/d2238e58f27cb3a1524c8432c04410ff4c0e25d3) or https://github.com/NeuroJSON/json/blob/issue3464/docs/mkdocs/docs/features/binary_formats/bjdata.md
> 
> not sure how to do a PR to a PR, feel free to copy the `bjdata.md` file directly to your local branch.

I think when creating a PR you can select a different destination branch than `develop`. This PR is being developed in the `documentation` branch (see line below the PR title).

**Comment by fangq** [comment]
> I think when creating a PR you can select a different destination branch than `develop`. This PR is being developed in the `documentation` branch (see line below the PR title).

thanks @falbrechtskirchinger, new PR against `documentation` branch at https://github.com/nlohmann/json/pull/3493

**Pull request merged**